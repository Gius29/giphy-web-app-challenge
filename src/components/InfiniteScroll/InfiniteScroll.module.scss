.infiniteScroll_one_column {
  list-style-type: none;
  padding-inline-start: 0;
}

.infiniteScroll_three_columns {
  list-style-type: none;
  padding-inline-start: 0;
  -webkit-column-count: 3;
  -moz-column-count: 3;
  column-count: 3;
}
