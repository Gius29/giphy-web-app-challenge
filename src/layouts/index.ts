// Each constant represent a possible layout that the app may have.
export enum LayoutIDs {
  ONE_COLUMN = "ONE_COLUMN",
  THREE_COLUMNS = "THREE_COLUMNS"
}
