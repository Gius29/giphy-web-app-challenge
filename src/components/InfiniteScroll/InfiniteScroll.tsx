import React, { PureComponent } from "react";
import { LayoutIDs } from "../../layouts";

import style from "./InfiniteScroll.module.scss";

interface StateProps {
  items: any;
  render: any;
  layoutID: LayoutIDs;
}

interface DispatchProps {
  searchQuery: () => void;
}

type InfiniteScrollProps = StateProps & DispatchProps;

class InfiniteScroll extends PureComponent<InfiniteScrollProps> {
  componentDidMount() {
    window.onscroll = () => {
      let bottomOfWindow =
        document.documentElement.scrollTop + window.innerHeight;
      if (bottomOfWindow === document.documentElement.offsetHeight) {
        this.props.searchQuery();
      }
    };
  }

  componentWillUnmount() {
    window.onscroll = null;
  }

  selectLayout() {
    const { layoutID } = this.props;
    if (layoutID === LayoutIDs.ONE_COLUMN) {
      return style.infiniteScroll_one_column;
    }
    // LayoutIDs.THREE_COLUMNS (There are no other layout possible)
    return style.infiniteScroll_three_columns;
  }

  render() {
    const { items, render } = this.props;
    const style = this.selectLayout();

    if (items && items.length > 0) {
      const l = items.map((item: any) => <li key={item.id}>{render(item)}</li>);
      return <ul className={style}>{l}</ul>;
    }

    return <ul />;
  }
}

export default InfiniteScroll;
