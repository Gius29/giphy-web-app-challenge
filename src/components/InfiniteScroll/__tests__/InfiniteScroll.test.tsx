import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import { LayoutIDs } from "../../../layouts";
import InfiniteScroll from "../InfiniteScroll";

describe("InfiniteScroll", () => {
  const mockItems = [1, 2, 3];

  it("should render a list of elements with a cloumn", () => {
    const wrapper = renderer.create(
      <InfiniteScroll
        items={mockItems}
        render={(item: number) => <div>{item}</div>}
        layoutID={LayoutIDs.ONE_COLUMN}
        searchQuery={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should render a list of elements with three columns", () => {
    const wrapper = renderer.create(
      <InfiniteScroll
        items={mockItems}
        render={(item: number) => <div>{item}</div>}
        layoutID={LayoutIDs.THREE_COLUMNS}
        searchQuery={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should return an empty list", () => {
    const wrapper = renderer.create(
      <InfiniteScroll
        items={[]}
        render={(item: number) => <div>{item}</div>}
        layoutID={LayoutIDs.THREE_COLUMNS}
        searchQuery={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should trigger a search query", () => {
    const mockSearchQuery = jest.fn();

    const wrapper = mount(
      <InfiniteScroll
        items={mockItems}
        render={(item: number) => <div>{item}</div>}
        layoutID={LayoutIDs.THREE_COLUMNS}
        searchQuery={mockSearchQuery}
      />
    );

    // Edit the values to trigger a pagination
    document.documentElement.scrollTop = 0;
    window.innerHeight = 0;

    window.onscroll();

    wrapper.unmount();
    expect(mockSearchQuery).toHaveBeenCalledTimes(1);
  });

  it("should trigger a search query", () => {
    const mockSearchQuery = jest.fn();

    const wrapper = mount(
      <InfiniteScroll
        items={mockItems}
        render={(item: number) => <div>{item}</div>}
        layoutID={LayoutIDs.THREE_COLUMNS}
        searchQuery={mockSearchQuery}
      />
    );

    // Edit the values to not trigger a pagination
    document.documentElement.scrollTop = 0;
    window.innerHeight = 100;

    window.onscroll();

    wrapper.unmount();
    expect(mockSearchQuery).toHaveBeenCalledTimes(0);
  });
});
