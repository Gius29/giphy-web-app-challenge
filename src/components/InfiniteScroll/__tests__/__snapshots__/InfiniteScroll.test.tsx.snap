// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteScroll should render a list of elements with a cloumn 1`] = `
<ul
  className="infiniteScroll_one_column"
>
  <li>
    <div>
      1
    </div>
  </li>
  <li>
    <div>
      2
    </div>
  </li>
  <li>
    <div>
      3
    </div>
  </li>
</ul>
`;

exports[`InfiniteScroll should render a list of elements with three columns 1`] = `
<ul
  className="infiniteScroll_three_columns"
>
  <li>
    <div>
      1
    </div>
  </li>
  <li>
    <div>
      2
    </div>
  </li>
  <li>
    <div>
      3
    </div>
  </li>
</ul>
`;

exports[`InfiniteScroll should return an empty list 1`] = `<ul />`;
